### PR TITLE
final commit for glance_image_cache

### PR DESCRIPTION
**Pull request by upadhyay-prashant**
final commit for caching image in SBS. 
for details please refer: JBS-66.

### CODE REVIEW

File: cinder/api/v2/cache_volumes.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    def get_cache_snapshot(self,req,body,volumeController):
+        self.volumeController = volumeController
+        volume=body['volume']
+
+        context=req.environ['cinder.context']
+        original_vol_name=volume['display_name']
+        self.update_context(context)
+        self.update_req_and_body(req,body)
+        snap=None
+        if 'imageRef' in volume and volume['imageRef'] is not None and  \
```
Comment (shishirng) [inline cinder/api/v2/cache_volumes.py]: if we are in get_cache_snapshot, the above condition is already satisfied

**Comment by shishirng** [comment]
+1. Looks good for me.

File: cinder/api/v2/volumes.py [hunk 2]
```diff
@@ -319,17 +331,59 @@ def create(self, req, body):
...
+            volume['display_name']=original_vol_name
 
+        if snapshot_id is not None:
+           try:
+               snap_from_cache=True
+               admin_context=context.elevated()
+               kwargs['snapshot'] = self.volume_api.get_snapshot(admin_context,snapshot_id)
+           except exception.NotFound:
+               volume['imageRef']=original_imageRef
+               snap_from_cache=False
```
Comment (aggarwalchirag) [inline cinder/api/v2/volumes.py]: Here, in case of exception, snapshot_id should be set to None

File: cinder/api/v2/volumes.py [hunk 3]
```diff
@@ -319,17 +331,59 @@ def create(self, req, body):
...
             volume['display_description'] = volume.get('description')
             del volume['description']
-
+        
+        snap_from_cache=False
+        original_imageRef=None
+        snapshot_id=None
         if 'image_id' in volume:
             volume['imageRef'] = volume.get('image_id')
+            original_imageRef=volume['imageRef']
```
Comment (aggarwalchirag) [inline cinder/api/v2/volumes.py]: original_imageref should be stored if either of image_id or image_ref is passed in volume. So, its better to have a separate if check later which checks for imageRef and correspondingly sets original_imageref

File: cinder/api/v2/volumes.py [hunk 4]
```diff
@@ -414,12 +459,17 @@ def create(self, req, body):
             if image_ref is not None:
                 image_uuid = self._image_uuid_from_ref(image_ref, context)
                 kwargs['image_id'] = image_uuid
```
Comment (aggarwalchirag) [inline cinder/api/v2/volumes.py]: we should not hit this piece of code, if we are creating a volume from cached snapshot

File: cinder/api/v2/cache_volumes.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                    snap = self.create_cache_snapshot()
+                return snap
+
+            snap = run_create_cache()
+
+        self.revert_req_and_body(context,original_vol_name)
+        if snap == None:
+            return None
+        else:
+            del volume['imageRef']
```
Comment (aggarwalchirag) [inline cinder/api/v2/cache_volumes.py]: in the final volume, do we give out the image id ?

File: cinder/api/v2/cache_volumes.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        self.cache_vol_name=None
+        self.req=None
+        self.body=None
+
+
+    def get_snap_cache(self):
+        snap_found=False
+        while not snap_found:
+            snap=self.volume_api.get_snapshot_by_name(self.context,self.cache_snap_name)
+            if 'status' in snap and snap['status']=="available":
```
Comment (aggarwalchirag) [inline cinder/api/v2/cache_volumes.py]: what if snap is none

File: cinder/api/v2/volumes.py (hunk 2 above)
Comment (upadhyay-prashant) [inline cinder/api/v2/volumes.py]: ideally this exception would never hit though for safety reverting snapshot id is necessary. welcome comment :)

**Pull request merged**